### ISSUE TITLE
services/horizon/expingest: Rewrite System to a state machine

### ISSUE THREAD

**Issue by bartekn** [stellar/go#1969]
`expingest.System` code is still easy to understand but after adding history ingestion code it will become much harder to follow and test.

I think we should rewrite `System` to a state machine. This will make it much easier to understand, test and will remove a need for `retry` functions. Each state will have it's own `System` method that will perform actions and change the state based on the action code and input params. The initial design can be found below.

### States and transitions

State can be one of the following:
* `Init` - Initial state that determines the next state based on the input values, like DB state, ingestion version, if current node is a leader or not, etc.
* `LoadStateInMemory` - Loads all the data that needs to be available in memory, like offers graph.
* `WaitForCheckpoint` - Waits for the next checkpoint in history archive. This supports one specific case when the last ledger in history table is ahead of history archive latest.
* `Catchup` - Runs `RangeSession` (https://github.com/stellar/go/pull/1890) to ingest history data up to the latest checkpoint in history archive.
* `BuildStateAndResume` - Clears state tables and runs `LiveSession` starting on a given ledger sequence, then continues ingesting using ledger backend.
* `Resume` - Resumes ingestion from the last ingested ledger.
* `Shutdown` - Does nothing, breaks main loop.

The main function that runs the machine will look like this:
```go
func run() {
	for {
		var err error
		var nextState State

		switch currentState {
		case Init:
			nextState, err := init()
		case LoadState:
			nextState, err := loadState()
		// ...
		case Shutdown:
			log.Info("Ingestion shut down")
			return
		default:
			panic("Unknown state!")
		}

		if err != nil {
			log.Error(err)
			time.Sleep(time.Second)
		}

		currentState = nextState
	}
}
```
#### `Init`

Here's a first draft of `Init` transition function:
```go
func init() (State, err) {
	// Here or as function params:
	ingestVersion := s.historyQ.GetExpIngestVersion()
	lastIngestedLedger := s.historyQ.GetLastLedgerExpIngest()
	lastHistoryLedger := s.historyQ.GetLastHistoryLedger()
	lastCheckpoint := historyArchive.GetLastCheckpoint()

	if lastHistoryLedger == 0 {
		// No ledgers in history_ledgers so start from the latest checkpoint and
		// continue using ledgerbackend.
		return BuildStateAndResume, nil
	}

	// Now: lastHistoryLedger != 0 => there are ledgers in history_ledgers

	if ingestVersion != CurrentVersion {
		// Ingestion version upgrade.
		switch {
		case lastHistoryLedger < lastCheckpoint-1:
			// Catchup to lastCheckpoint-1 and then BuildStateAndResume
			return Catchup, nil
		case lastHistoryLedger == lastCheckpoint-1:
			// We can BuildStateAndResume without catching up
			return BuildStateAndResume, nil
		case lastHistoryLedger > lastCheckpoint-1:
			// Wait for the next checkpoint that is ahead of lastHistoryLedger
			return WaitForCheckpoint, nil
		}
	}

	// Now:
	// - ingestVersion == CurrentVersion => no upgrade
	// - lastHistoryLedger != 0 => there are ledgers in history_ledgers
	return LoadStateInMemory, nil
}
```
#### `LoadStateInMemory`

Works exactly like current `loadOrderBookGraphFromDB` and changes the state to `Resume`.

#### `WaitForCheckpoint`

This state is to support waiting for history checkpoint that will be ahead of the last ingested ledger and changes the state to `Catchup`.

#### `Catchup`

Catchup ingest historical data without processing meta (ledger entries) up to a given checkpoint ledger and changes the state to `BuildStateAndResume`.

#### `BuildStateAndResume`

Clears state tables, update key value store, then starts `LiveSession` (ingests state using given checkpoint and follow ledgers using `ledgerbackend`). Changes the state to `BuildStateAndResume` in case of errors ingesting state, `Resume` in case of errors ingesting ledgers or `Shutdown` (when Horizon calls `System.Shutdown()` that is passed to session).

#### `Resume`

Works like current `resumeFromLedger`. Changes the state to `Resume` or `Shutdown` (when Horizon calls `System.Shutdown()` that is passed to session).

### State diagram

```dot
digraph StateTransitions {
  "Init" -> "BuildStateAndResume"
  "Init" -> "Catchup"
  "Init" -> "WaitForCheckpoint"
  "Init" -> "LoadStateInMemory"
  "LoadStateInMemory" -> "Resume"
  "WaitForCheckpoint" -> "Catchup"
  "Catchup" -> "BuildStateAndResume"
  "BuildStateAndResume" -> "BuildStateAndResume"
  "BuildStateAndResume" -> "Resume"
  "BuildStateAndResume" -> "Shutdown"
  "Resume" -> "Resume"
  "Resume" -> "Shutdown"
}
```

![Screenshot 2019-11-21 at 15 39 56](https://user-images.githubusercontent.com/464938/69347673-351cac80-0c75-11ea-9ef9-b1c61d9ada2d.png)

### Possible improvements

The machine above is the initial version that can be improved and optimized. For example: If a difference between the latest history checkpoint and latest ledger in `history_ledger` is big, it's possible to `Catchup` (without ingesting meta) and then `BuildStateAndResume` that can be faster than ingesting meta for each ledger.

**Comment by MonsieurNicolas** [comment]
I think moving to state machines makes total sense. I am maybe just confused by the names: I would expect the system to basically "build an initial state" (which could be "no op") and then "ingest ledgers" (regardless of using meta or not), but I don't really see this in this diagram as there are extra states and transitions that I don't grok with what you described here.

**Comment by bartekn** [comment]
Closed in #2085.
